### PR TITLE
feat(j-s): Allow multiple subtypes in indictments

### PR DESCRIPTION
**Pull request by oddsson**
# Allow multiple subtypes in indictments

[Asana](https://app.asana.com/0/1199153462262248/1208795316278077/f)

## What

The changes in this PR enable users to create indictments with multiple subtypes. The main complexity with this change is 

1. Users can create indictment counts with one or more subtypes. Furthermore, if an indictment count has multiple subtypes, users can select individual indictment counts to work with.
2. The indictment description should be autofilled with appropriate text based on what subtypes an indictment count is created with and what subtype is selected in the indictment count. 
3. The `requestDriversLicenseSuspension` and demands should be correctly autofilled based on whether or not there are traffic violation indictment counts.
4. The Indictment screen (prev TrafficViolation screen) should be correctly validated. 
5. Please refer to the Asana task to see what exactly should happen when certain subtypes are selected.

There are also other, smaller changes implemented

- Fix a typo in one of our Slack messages: Beðið um aftuköllun ✍️ Beðið um aftu**r**köllun
- Rename `stepHelper` file in `utils` to `utils`. `stepHelper` is a bit outdated and `utils` is a bit clearer IMO. This change is responsible for most of the files changed in this PR.

## Why

This is crucial for us to be able to offer combo indictments.

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

## Release Notes

- **Refactoring**
	- Consolidated utility functions from `stepHelper` to `utils` module across multiple components.
	- Updated import paths for various utility functions like `hasSentNotification`, `isBusiness`, and `shouldUseAppealWithdrawnRoutes`.

- **Localization**
	- Added new localization messages for indictment and substances-related fields.
	- Enhanced titles and labels for legal process input fields.

- **Indictment Handling**
	- Improved validation and handling of indictment counts.
	- Added new utility functions for traffic violation and indictment count processing.
	- Enhanced logic for managing police case information and subtypes.

- **Testing**
	- Expanded test coverage for utility functions and incident descriptions.
	- Added new test scenarios for handling various indictment subtypes.

- **Minor Improvements**
	- Updated function signatures to support more flexible parameter handling.
	- Simplified some code structures using shorthand property names.

### CODE REVIEW

File: apps/judicial-system/web/src/utils/validate.ts [hunk 1]
```diff
@@ -298,8 +298,14 @@ export const isTrafficViolationStepValidIndictments = (
...
-    workingCase.demands &&
-      (!workingCase.hasCivilClaims || workingCase.civilDemands),
+    hasOnlyOneItemInSubArrays(workingCase.indictmentSubtypes) ||
+      (workingCase.indictmentCounts?.every(
+        (indictmentCount) =>
+          indictmentCount.indictmentCountSubtypes &&
+          indictmentCount.indictmentCountSubtypes?.length > 0,
+      ) &&
+        workingCase.demands &&
+        (!workingCase.hasCivilClaims || workingCase.civilDemands)),
```
[the remark this answers is not in the document]
Comment (unakb) [reply apps/judicial-system/web/src/utils/validate.ts:308]: I agree with this, I find the logic a bit hard to follow in the current implementation.

File: apps/judicial-system/web/src/utils/validate.ts [hunk 2]
```diff
@@ -298,8 +298,14 @@ export const isTrafficViolationStepValidIndictments = (
...
   return Boolean(
-    workingCase.demands &&
-      (!workingCase.hasCivilClaims || workingCase.civilDemands),
+    hasOnlyOneItemInSubArrays(workingCase.indictmentSubtypes) ||
+      (workingCase.indictmentCounts?.every(
+        (indictmentCount) =>
+          indictmentCount.indictmentCountSubtypes &&
+          indictmentCount.indictmentCountSubtypes?.length > 0,
+      ) &&
+        workingCase.demands &&
```
Comment (unakb) [inline apps/judicial-system/web/src/utils/validate.ts:307]: Did you mean to stop doing the workingCase.demands checks if hasOnlyOneItemInSubArrays returns true?

File: apps/judicial-system/web/src/utils/utils.ts [hunk 3]
```diff
@@ -147,3 +147,30 @@ export const shouldUseAppealWithdrawnRoutes = (theCase: Case): boolean => {
...
+ * }
+ *
+ * const result = hasOnlyOneItemInSubArrays(indictmentSubtypes)
+ * console.log(result) // true
+ *
+ * @param indictmentSubtypes - The object containing the subarrays to check.
+ * @returns {boolean} - True if all subarrays have only one item, false otherwise.
+ */
+export const hasOnlyOneItemInSubArrays = (
+  indictmentSubtypes?: Record<string, string[]>,
```
Comment (unakb) [inline apps/judicial-system/web/src/utils/utils.ts:167]: Consider naming this something other than indictmentSubtypes as the rest of the function seems to be more general, so we could theoretically use it to check this for other subarrays that aren't indictment subtypes

File: apps/judicial-system/web/src/routes/CourtOfAppeal/Summary/Summary.spec.tsx [hunk 4]
```diff
@@ -9,11 +9,11 @@ import {
   CaseType,
 } from '@island.is/judicial-system-web/src/graphql/schema'
 import { mockTransitonCaseMutation } from '@island.is/judicial-system-web/src/utils/mocks'
-import * as stepHelper from '@island.is/judicial-system-web/src/utils/stepHelper'
 import {
   FormContextWrapper,
   IntlProviderWrapper,
 } from '@island.is/judicial-system-web/src/utils/testHelpers'
+import * as stepHelper from '@island.is/judicial-system-web/src/utils/utils'
```
[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/routes/CourtOfAppeal/Summary/Summary.spec.tsx]: The only file with a `stepHelper` reference was `apps/judicial-system/web/src/routes/CourtOfAppeal/Summary/Summary.spec.tsx` and that has now been fixed.

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/PoliceCaseInfo/PoliceCaseInfo.tsx [hunk 5]
```diff
@@ -209,9 +215,12 @@ export const PoliceCaseInfo: FC<Props> = ({
           placeholder={formatMessage(policeCaseInfo.indictmentTypePlaceholder)}
           onChange={(selectedOption) => {
             const indictmentSubtype = selectedOption?.value as IndictmentSubtype
+
             updatePoliceCase(index, {
-              subtypes: [...(subtypes || []), indictmentSubtype],
+              subtypes: [...subtypesArray, indictmentSubtype],
             })
+
+            updateIndictmentCount(policeCaseNumbers[index], crimeScene || {})
```
[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/PoliceCaseInfo/PoliceCaseInfo.tsx]: Turns out, I don't need to call this function here. I've removed it.

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Indictment/IndictmentCount.tsx [hunk 6]
```diff
@@ -287,41 +289,77 @@ export const getIncidentDescription = (
...
+    : '[Dagsetning]'
+
+  const vehicleRegistration =
+    vehicleRegistrationNumber || '[Skráningarnúmer ökutækis]'
+
+  const hasSingleSubtype = hasOnlyOneItemInSubArrays(subtypes)
+
+  const singleSubType =
+    (policeCaseNumber && subtypes?.[policeCaseNumber]?.[0]) || undefined
+
```
[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/routes/Prosecutor/Indictments/Indictment/IndictmentCount.tsx:314]: Am I not doing that by writing `subtypes?.`?

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Indictment/IndictmentCount.tsx [hunk 7]
```diff
@@ -469,300 +516,339 @@
...
+          </>
+        )}
       </Box>
-      {subtypes.length > 1 && (
-        <Box marginBottom={2}>
+      {(isTrafficViolationCase(workingCase) ||
+        (indictmentCount?.indictmentCountSubtypes?.includes(
+          IndictmentSubtype.TRAFFIC_VIOLATION,
+        ) ??
+          false)) && (
```
[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/routes/Prosecutor/Indictments/Indictment/IndictmentCount.tsx:568]: This code is as it should be. It checks if the case is a pure traffic violation case, i.e. the only subtype is traffic violation or if the individual indictment count has a traffic violation selected.

File: apps/judicial-system/web/src/utils/utils.ts (hunk 3 above)
Comment (oddsson) [reply apps/judicial-system/web/src/utils/utils.ts]: Yes, thanks. I've renamed the parameter and some other properties of this function. Let me know if anything is unclear.

File: apps/judicial-system/web/src/utils/validate.ts (hunk 2 above)
Comment (oddsson) [reply apps/judicial-system/web/src/utils/validate.ts]: Yes, and that is clearer now that I've refactored the code. Please take another look and report back if anything is unclear :)

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Indictment/IndictmentCount.tsx [hunk 8]
```diff
@@ -469,300 +516,339 @@
...
+                  autoComplete="off"
+                  label={formatMessage(strings.bloodAlcoholContentLabel)}
+                  placeholder={formatMessage(
+                    strings.bloodAlcoholContentPlaceholder,
+                  )}
+                  errorMessage={bloodAlcoholContentErrorMessage}
+                  hasError={bloodAlcoholContentErrorMessage !== ''}
+                  required
+                />
+              </InputMask>
```
[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/routes/Prosecutor/Indictments/Indictment/IndictmentCount.tsx:758]: No thanks..

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Indictment/IndictmentCount.tsx [hunk 9]
```diff
@@ -424,6 +463,9 @@ export const IndictmentCount: FC<Props> = ({
 
     handleIndictmentCountChanges({
       indictmentCountSubtypes: Array.from(currentSubtypes),
+      ...(!currentSubtypes.has(IndictmentSubtype.TRAFFIC_VIOLATION) && {
+        offenses: [],
+      }),
```
[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/routes/Prosecutor/Indictments/Indictment/IndictmentCount.tsx:468]: No thanks, I'm good.

[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/routes/Prosecutor/Indictments/Indictment/IndictmentCount.tsx]: Yes! Thanks, done

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Indictment/IndictmentCount.tsx [hunk 10]
```diff
@@ -469,300 +518,339 @@ export const IndictmentCount: FC<Props> = ({
...
+            <InputMask
+              mask={[/[A-Z]/i, /[A-Z]/i, /[A-Z]|[0-9]/i, /[0-9]/, /[0-9]/]}
+              maskPlaceholder={null}
+              value={indictmentCount.vehicleRegistrationNumber ?? ''}
+              beforeMaskedStateChange={({ nextState }) => {
+                let { value } = nextState
+                value = value.toUpperCase()
+
+                return { ...nextState, value }
+              }}
```
[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/routes/Prosecutor/Indictments/Indictment/IndictmentCount.tsx:587]: No, thanks. It's on the roadmap to remove the constraints on the vehicle registration number so we're not changing it for now.

File: apps/judicial-system/web/src/utils/validate.ts [hunk 11]
```diff
@@ -1,5 +1,6 @@
 // TODO: Add tests
 import {
+  IndictmentSubtype,
```
[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/utils/validate.ts:5]: Nah, thanks. The partition function is much more readable than this.

File: apps/judicial-system/web/src/utils/validate.ts [hunk 12]
```diff
@@ -297,10 +302,65 @@ export const isProcessingStepValidIndictments = (
...
+
+  const validateIndictmentCount = (indictmentCount: IndictmentCount) =>
+    indictmentCount.policeCaseNumber &&
+    indictmentCount.offenses &&
+    indictmentCount.offenses?.length > 0 &&
+    indictmentCount.vehicleRegistrationNumber &&
+    indictmentCount.lawsBroken &&
+    indictmentCount.incidentDescription &&
+    indictmentCount.legalArguments
+
```
[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/utils/validate.ts:322]: No thanks, I think it's clear enough.

File: apps/judicial-system/web/src/utils/validate.ts [hunk 13]
```diff
@@ -297,10 +301,72 @@ export const isProcessingStepValidIndictments = (
...
+  ].every(validateIndictmentCount)
+
+  const isOtherIndictmentCountsValid =
+    nonTrafficViolationIndictmentCounts.every(
+      (indictmentCount) =>
+        indictmentCount.indictmentCountSubtypes &&
+        indictmentCount.indictmentCountSubtypes.length > 0,
+    )
+
+  return isTrafficViolationIndictmentCountsValid && isOtherIndictmentCountsValid
```
[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/utils/validate.ts]: Done, thanks.

File: apps/judicial-system/web/src/utils/validate.ts [hunk 14]
```diff
@@ -297,10 +301,74 @@
...
```
[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/utils/validate.ts]: Done, I've added tests for the validation. Thanks.

File: apps/judicial-system/web/src/utils/validate.ts [hunk 15]
```diff
@@ -297,10 +300,53 @@ export const isProcessingStepValidIndictments = (
...
+  }
+
+  const hasValidIndictmentCounts =
+    workingCase.indictmentCounts?.every(
+      (indictmentCount) =>
+        indictmentCount.indictmentCountSubtypes &&
+        indictmentCount.indictmentCountSubtypes.length > 0,
+    ) ?? false
+
+  return hasValidIndictmentCounts && hasValidDemands
```
[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/utils/validate.ts]: I think I'm good. I know the function is a bit complex but I feel that your solution doesn't make it that much simpler.

File: apps/judicial-system/web/src/utils/formHelper.ts [hunk 16]
```diff
@@ -317,3 +321,22 @@ export const findFirstInvalidStep = (steps: string[], theCase: Case) => {
...
+  policeCaseNumber?: string | null,
+  indictmentSubtypes?: IndictmentSubtypeMap,
+) => {
+  if (!policeCaseNumber || !indictmentSubtypes) {
+    return false
+  }
+
+  if (
+    indictmentSubtypes[policeCaseNumber].length === 1 &&
+    indictmentSubtypes[policeCaseNumber][0] ===
```
Comment (thorhildurt) [inline apps/judicial-system/web/src/utils/formHelper.ts:335]: Is it guaranteed that the subtype `TRAFFIC_VIOLATION` will always be the first instance of the array?

File: apps/judicial-system/web/src/utils/validate.ts [hunk 17]
```diff
@@ -297,10 +302,65 @@ export const isProcessingStepValidIndictments = (
...
+  const hasValidDemands = Boolean(
     workingCase.demands &&
       (!workingCase.hasCivilClaims || workingCase.civilDemands),
   )
+
+  if (!workingCase.indictmentSubtypes || !hasValidDemands) {
+    return false
+  }
+
+  const validateIndictmentCount = (indictmentCount: IndictmentCount) =>
```
Comment (thorhildurt) [inline apps/judicial-system/web/src/utils/validate.ts:314]: Niiit: since this returns a boolean what do you think about referring to this as `isValidIndictmentCount` instead of `validateIndictmentCount`? Since `validateIndictmentCount` might indicate a simple `void` func for example?

File: apps/judicial-system/web/src/utils/formHelper.ts (hunk 16 above)
Comment (thorhildurt) [reply apps/judicial-system/web/src/utils/formHelper.ts:335]: AH we check if the array is only has exactly one element 👍 🙈

File: apps/judicial-system/web/src/utils/validate.ts [hunk 18]
```diff
@@ -297,10 +302,65 @@ export const isProcessingStepValidIndictments = (
...
+  }
+
+  // Some indictment counts are traffic violations while others are not
+  const [onlyTrafficViolationIndictmentCounts, otherIndictmentCounts] =
+    partition(workingCase.indictmentCounts, (indictmentCount) => {
+      const policeCaseNumber = indictmentCount.policeCaseNumber
+
+      return isTrafficViolationIndictmentCount(
+        policeCaseNumber,
+        workingCase.indictmentSubtypes,
```
Comment (thorhildurt) [inline apps/judicial-system/web/src/utils/validate.ts:338]: What if we have multiple subtypes here and one of them is `IndictmentSubtype.TRAFFIC_VIOLATION`, then `isTrafficViolationIndictmentCount` will return `false`, right? Is that expected here?

**Review by thorhildurt — COMMENTED** [review]
Nice, left some comments. Mostly questions and non-blocking nits ☺️

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/PoliceCaseInfo/PoliceCaseInfo.tsx [hunk 19]
```diff
@@ -82,6 +86,8 @@ export const PoliceCaseInfo: FC<Props> = ({
   const [policeCaseNumberErrorMessage, setPoliceCaseNumberErrorMessage] =
     useState<string>('')
 
+  const subtypesArray = [...(subtypes || [])]
```
[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/PoliceCaseInfo/PoliceCaseInfo.tsx]: Yes 🙈 Thanks

File: apps/judicial-system/web/src/utils/validate.ts (hunk 17 above)
Comment (oddsson) [reply apps/judicial-system/web/src/utils/validate.ts]: Yes, done ✔️ Thanks..

File: apps/judicial-system/web/src/utils/validate.ts (hunk 18 above)
Comment (oddsson) [reply apps/judicial-system/web/src/utils/validate.ts]: This comment was made before a refactor was done on this code. Could you please take another look 👀🙏

File: apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/PoliceCaseInfo/PoliceCaseInfo.tsx [hunk 20]
```diff
@@ -231,8 +242,21 @@ export const PoliceCaseInfo: FC<Props> = ({
...
                   })
+
+                  updateIndictmentCount(
+                    policeCaseNumbers[index],
+                    crimeScene || {},
+                    {
+                      [policeCaseNumbers[index]]: subtypes.filter(
+                        (s) =>
+                          s !== subtype &&
+                          indictmentCount?.indictmentCountSubtypes?.includes(s),
```
[the remark this answers is not in the document]
Comment (oddsson) [reply apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/PoliceCaseInfo/PoliceCaseInfo.tsx:256]: Sure, let's discuss this IRL 🤝

[the remark this answers is not in the document]
Comment (thorhildurt) [reply apps/judicial-system/web/src/routes/Prosecutor/Indictments/Defendant/PoliceCaseInfo/PoliceCaseInfo.tsx:256]: I actually get it now. So these would be the selected subtypes for a given indictment count. While the subtypes on the working case, are all subtypes on the case 👍

**Review by unakb — APPROVED** [review]

**Review by thorhildurt — APPROVED** [review]
